### ISSUE TITLE
RUF052: Don't flag assignments to private function arguments

### ISSUE THREAD

**Issue by aiudirog** [astral-sh/ruff#14968]
This is a follow up to #14796: attempting to assign a private function parameter still gets flagged in Ruff v0.8.3 ([playground](https://play.ruff.rs/e233076b-9184-492f-aa36-94306356bf12))

My primary use case for private parameters is when I need to track what I've seen when recursing through a potentially circular graph:

```python
class Node:

    connected: list[Node]

    def recurse(self, *, _seen: set[Node] | None = None):
        if _seen is None:
            _seen = set()  # Local dummy variable `_seen` is accessed (RUF052)
        elif self in _seen:
            return
        _seen.add(self)
        for other in self.connected:
            other.recurse(_seen=_seen)
```

In most cases, I wouldn't want the caller to have to seed the `_seen` param or even know that it exists, as imo, that's just an implementation detail.

**Comment by dylwil3** [comment]
Incidentally, in this example the rule is _not_ complaining about the binding for the function parameter, but rather about the rebinding in the if-statement. So I think we need to add some logic that skips if the offending binding is a re-binding of a private parameter declaration.

Unrelated - would it be more helpful if the diagnostic pointed to the problematic _reference_ rather than the _binding_?

**Comment by aiudirog** [comment]
> Incidentally, in this example the rule is _not_ complaining about the binding for the function parameter, but rather about the rebinding in the if-statement. So I think we need to add some logic that skips if the offending binding is a re-binding of a private parameter declaration.

Yeah on 0.8.2 the rule flagged both the function definition and the reassignment in the body separately, so that seems like the appropriate solution.

> Unrelated - would it be more helpful if the diagnostic pointed to the problematic _use_ rather than the _binding_?

That might just flip it to error on `_seen.add(self)` and, if there's a lot of usages in the same function, could output a lot of messages
